### ISSUE TITLE
perf(morphology): add an exact, low-memory "shift" engine and pick the engine by kernel size

### ISSUE THREAD

**Issue by ducha-aiki** [kornia/kornia#4729]
## Proposal

Add a third engine, `engine="shift"`, to `kornia.morphology` (`dilation`, `erosion`, and the compound ops through them). It builds no window tensor. It loops over the k² kernel positions, takes the padded image shifted by each offset as a view, and keeps a running elementwise max (dilation) or min (erosion):

```python
out = padded[..., 0:H, 0:W] + offsets[0, 0]
for i in range(kh):
    for j in range(kw):
        v = padded[..., i:i+H, j:j+W] + offsets[i, j]
        out = torch.maximum(out, v)          # torch.minimum for erosion
```

Then have `engine="auto"` (#4728) pick an engine by kernel size (`kh * kw`) as well as by device, instead of by device alone.

Why:

- **Exact.** It evaluates the same max-plus expression as `unfold`, in a different order, so it has no backend-precision problem. `convolution` inherits cuDNN TF32 and macOS float32 CPU rounding (up to ~4e-3 measured on an M1).
- **Low memory.** Its intermediate is B·C·H·W, not the B·C·H·W·k² that `unfold` and `convolution` both build (#2948 / #3001).
- **Compile-friendly.** Inductor fuses the whole loop into one kernel. Compiled `convolution` stays an opaque conv followed by a reduction, and on MPS it is 5–14x slower than compiled `unfold` (#4728, open question 1).
- **No `is_compiling()` gate.** It removes the need for #4728 to choose one engine for eager and another for compile.

## Reference implementation (verified)

It is a drop-in change inside `dilation`/`erosion` after the existing padding and `neighborhood` code; nothing else changes:

```python
def _shift_reduce(padded: torch.Tensor, offsets: torch.Tensor, h: int, w: int, dilate: bool) -> torch.Tensor:
    kh, kw = offsets.shape
    out = padded[..., 0:h, 0:w] + offsets[0, 0]
    for i in range(kh):
        for j in range(kw):
            if i == 0 and j == 0:
                continue
            v = padded[..., i : i + h, j : j + w] + offsets[i, j]
            out = torch.maximum(out, v) if dilate else torch.minimum(out, v)
    return out

# dilation:  _shift_reduce(padded, neighborhood.flip((0, 1)).to(tensor.dtype), H, W, dilate=True)
# erosion:   _shift_reduce(padded, (-neighborhood).to(tensor.dtype), H, W, dilate=False)
```

Checked against `engine="unfold"` on `main` (d0b8328e2) with `torch.equal`, not `allclose`:

- 720 cases, **0 mismatches**, both ops: CPU float32/float16/bfloat16/float64 and MPS float32/float16; kernels 1×1, 3×3, 3×5, 4×2 and 7×7 with random zeros; default, `[0, 0]` and bottom-right origins; `geodesic` and `constant` (`border_value=0.5`) borders; with and without a random non-flat `structuring_element`.
- `reflect`, `replicate` and `circular` borders: bitwise equal on CPU and MPS.
- Gradients: `gradcheck` passes for both ops in float64, and the gradient is bitwise equal to `unfold`'s on tie-free input. Where maxima tie, each engine sends the gradient to one of the tied pixels, and not always the same one. That is already true between `unfold` and `convolution`.
- Not yet checked: TorchScript (`test_jit`), with the Python loop over `kernel.shape`; ONNX export.

## Measurements

Apple M1, torch 2.14.0, 4 threads, `dilation`, 8×3×256×256 float32, square all-ones kernel, `torch.utils.benchmark` median (ms). `conv` on CPU at k ≥ 15 was skipped (6.5 s at k15).

| device | k | eager unfold | eager conv | eager shift | compiled unfold | compiled shift | compile time unfold / shift (s) |
|---|--:|--:|--:|--:|--:|--:|--:|
| CPU | 3 | 24.6 | 17.0 | **9.7** | 2.66 | 2.73 | 1.9 / 0.2 |
| CPU | 5 | 55.4 | 108.9 | **23.4** | 17.2 | **5.95** | 1.3 / 2.3 |
| CPU | 7 | 76.0 | 220.3 | **49.3** | 18.8 | **11.7** | 0.1 / 0.4 |
| CPU | 9 | 98.5 | 391.1 | **80.6** | 33.1 | **19.4** | 1.0 / 5.9 |
| CPU | 11 | 133.5 | 747.4 | **120.3** | 41.6 | **32.7** | 1.0 / 8.6 |
| CPU | 15 | **199.2** | — | 218.7 | 72.3 | **62.6** | 1.1 / 1.6 |
| CPU | 21 | **362.4** | — | 437.2 | 144.6 | **123.6** | 1.1 / 32.4 |
| MPS | 3 | 179.3 | 12.0 | **12.0** | 1.39 | **1.03** | 0.1 / 0.1 |
| MPS | 5 | 306.3 | **21.5** | 30.7 | 1.87 | **1.39** | 0.5 / 0.9 |
| MPS | 7 | 445.2 | **45.9** | 58.4 | 2.83 | **1.53** | 0.1 / 0.6 |
| MPS | 9 | 591.2 | 97.2 | **95.5** | 37.3 | **2.86** | 0.3 / 2.7 |
| MPS | 11 | 751.8 | 179.1 | **140.9** | 44.9 | **5.17** | 0.3 / 4.7 |
| MPS | 15 | 1104.8 | 571.9 | **257.6** | 60.6 | **15.0** | 0.2 / 1.6 |
| MPS | 21 | 1464.1 | 2024.0 | **498.1** | 96.4 | **32.3** | 0.8 / 19.5 |

Compile times are single wall-clock measurements, and some rows hit a warm inductor cache (k15), so read them only as a trend: `shift` compile time grows with k², reaching 20–32 s at 21×21.

## How it would tie to kernel size (for the maintainer to confirm on CUDA and x86)

- **CPU, eager:** `shift` wins up to 11×11 (2.5x over `unfold` at 3×3, 1.1x at 11×11), and `unfold` wins from 15×15. That suggests `shift` when `kh * kw <= 121`, `unfold` above.
- **MPS, eager:** `shift` ties `convolution` at 3×3, loses 1.3–1.4x at 5×5–7×7, and wins from 9×9 (4x at 21×21). Both are far ahead of `unfold`. `shift` for every size costs at most 1.4x against `convolution` and avoids its precision caveat.
- **Compiled:** `shift` is as fast or faster everywhere, 13x on MPS at 9×9. The only cost is compile time at large kernels, which suggests the same k² cap, or accepting it.
- **Unmeasured:** CUDA eager, where 2·k² small launches may cost more than on MPS; x86 CPU; float16/bfloat16 timings; backward timings. `benchmarks/morphology/engines.py` in #4728 already runs the unfold/conv A/B with `--compile`, `--backward` and a CUDA TF32 column; adding `"shift"` to its `ENGINES` tuple covers the new engine.

A possible `auto` rule, pending those numbers:

```python
if engine == "auto":
    engine = "shift" if kh * kw <= 121 else ("unfold" if tensor.device.type == "cpu" else "convolution")
```

For kernels above 121 elements on non-CPU devices, the MPS data would favour `shift` over `convolution` as well; CUDA should decide that branch.

## Suggested tests

- A parametrized `torch.equal(engine="shift", engine="unfold")` over the matrix above (origins, borders, structuring element, dtypes).
- `gradcheck` for the new engine.
- `test_jit` with `engine="shift"`.
- A `fullgraph=True` compile test.

Related: #4525, #4728, #2948 / #3001.

Posted on behalf of @ducha-aiki by Claude (Claude Opus 5).
